### ISSUE TITLE
Support for a virtual filesystem in the style of `@yao-pkg/pkg` for a single file executable

### ISSUE THREAD

**Issue by segevfiner** [oven-sh/bun#9392]
### What is the problem this feature would solve?

Many Node.js modules use the filesystem and require some files alongside them to function. Some might be some static assets, and some might use native node modules or shared libraries. pkg handles this by trying to heuristically detect those, with a set of definitions of how to handle specific modules and the ability for the user to specify additional files to include. It then monkey patches the Node.js `fs` module to create a virtual folder to serve those files, and also handles extracting native node addon modules to disk along with their dependent shared libs so they can be loaded.

### What is the feature you are proposing to solve the problem?

Support such a virtual file system mechanism in bun, whether bun wants to also handle the complexity of figure out what to include in such a virtual fs for a single file executable, or not, or to leave that to a separate project is up to you.

**Comment by Jarred-Sumner** [comment]
You can embed files in `bun build --compile` if you explicitly import them.

We have a note about this [in the docs](https://bun.sh/docs/bundler/executables#embedding-files)

**Comment by segevfiner** [comment]
The question is how you will get it to work with arbitrary npm packages that use Node.js `fs` to read files from inside the package, like `pkg` handles.

**Comment by segevfiner** [comment]
Yes. In my own code, but what about third party modules that just `path.join` from `__dirname`. Like virtually the entire npm ecosystem, how can we handle those? That's why `pkg` has its virtual filesystem thing...

**Comment by meoyawn** [comment]
Deno supports this through an explicit [self-extracting compile mode](https://docs.deno.com/runtime/reference/cli/compile/#self-extracting-executables): embedded files materialize on first run, then later runs reuse them, with documented [startup and disk trade-offs](https://docs.deno.com/runtime/reference/cli/compile/#trade-offs). Bun could offer the same opt-in shape via a --self-extracting flag on bun build --compile, preserving ordinary package layouts for native addons without package-specific extraction shims.
